pkg/gofr/datasource/opentsdb/annotation.go

```diff
@@ -0,0 +1,334 @@
+	// StartTime is the Unix epoch timestamp (in seconds) for when the event occurred. This is required.
+	StartTime int64 `json:"startTime,omitempty"`
+
+	// EndTime is the optional Unix epoch timestamp (in seconds) for when the event ended, if applicable.
+	EndTime int64 `json:"endTime,omitempty"`
+
+	// Tsuid is the optional time series identifier if the annotation is linked to a specific time series.
+	Tsuid string `json:"tsuid,omitempty"`
+
+	// Description is a brief, optional summary of the event (recommended to keep under 25 characters for display purposes).
+	Description string `json:"description,omitempty"`
+
+	// Notes is an optional, detailed description of the event.
+	Notes string `json:"notes,omitempty"`
+
+	// Custom is an optional key/value map to store any additional fields and their values.
+	Custom map[string]string `json:"custom,omitempty"`
+}
+
+// AnnotationResponse encapsulates the response data and status when interacting with the /api/annotation endpoint.
+type AnnotationResponse struct {
+	// StatusCode holds the HTTP status code for the annotation request.
+	StatusCode int
+
+	// Annotation holds the associated annotation object.
+	Annotation
+
+	// ErrorInfo contains details about any errors that occurred during the request.
+	ErrorInfo map[string]interface{} `json:"error,omitempty"`
+
+	logger Logger
+	tracer trace.Tracer
+	ctx    context.Context
+}
+
+func (annotResp *AnnotationResponse) SetStatus(code int) {
+	setStatus(annotResp.ctx, annotResp, code, "SetStatus-Annotation", annotResp.logger)
+}
+
+func (annotResp *AnnotationResponse) setStatusCode(code int) {
+	annotResp.StatusCode = code
+}
+
+func (annotResp *AnnotationResponse) GetCustomParser() func(respCnt []byte) error {
+	return getCustomParser(annotResp.ctx, annotResp, "GetCustomParser-Annotation", annotResp.logger,
+		func(resp []byte) error {
+			originContents := string(resp)
+
+			var resultBytes []byte
+
+			if strings.Contains(originContents, "startTime") || strings.Contains(originContents, "error") {
+				resultBytes = resp
+			} else if annotResp.StatusCode == http.StatusNoContent {
+				return nil
+			}
+
+			return json.Unmarshal(resultBytes, &annotResp)
+		})
+}
+
+func (annotResp *AnnotationResponse) String() string {
+	return toString(annotResp.ctx, annotResp, "ToString-Annotation", annotResp.logger)
+}
+
+func (c *OpentsdbClient) QueryAnnotation(queryAnnoParam map[string]interface{}) (*AnnotationResponse, error) {
+	span := c.addTrace(c.ctx, "QueryAnnotation")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "QueryAnnotation", &status, &message, span)
+
+	if len(queryAnnoParam) == 0 {
+		message = "annotation query parameter is empty"
+		return nil, errors.New(message)
+	}
+
+	buffer := bytes.NewBuffer(nil)
+
+	size := len(queryAnnoParam)
+
+	i := 0
+
+	for k, v := range queryAnnoParam {
+		fmt.Fprintf(buffer, "%s=%v", k, v)
+
+		if i < size-1 {
+			buffer.WriteString("&")
+		}
+
+		i++
+	}
```

Comment:
> I'm a bit worried about how things could need to be escaped
> 
> Like an equal sign or an ampersand.
> 

Can you explain more ? .....I am not able to get you. I understand the url.Values part, but its just adding more complexity. I can add a comment saying that this is to prepare the url something of that sort, but I feel that is also not needed.

Reply:
My problem is the fact the following characters would lead to oddities

& ? #

If you are manipulating things that are like URL parameters, you should use url.Values

But once again, I'm unsure what you are trying to achieve here, or what are the annotations about